src/DebugTrait.php

```diff
@@ -14,74 +14,121 @@
+    {
+        if (isset($this->app) && $this->app instanceof \Psr\Log\LoggerInterface) {
+            $this->app->log($level, $message, $context);
+        } else {
+            $this->_echo_stderr("$message\n");
+        }
+
+        return $this;
+    }
 
-    public function debugTraceChange($trace = 'default')
+    /**
+     * Output message that needs to be acknowledged by application user. Make sure
+     * that $context does not contain any sensitive information.
+     *
+     * @param string $message
+     * @param array  $context
+     *
+     * @return $this
+     */
+    public function userMessage($message, $context = [])
     {
-        if ($this->isDebugEnabled()) {
-            $bt = [];
-            foreach (debug_backtrace() as $line) {
-                if (isset($line['file'])) {
-                    $bt[] = $line['file'].':'.$line['line'];
-                }
-            }
+        if (isset($this->app) && $this->app instanceof \atk4\core\AppUserNotificationInterface) {
+            $this->app->userNotification($message, $context);
+        } elseif (isset($this->app) && $this->app instanceof \Psr\Log\LoggerInterface) {
+            $this->app->log('warning', 'Could not notify user about: '.$message, $context);
+        } else {
+            $this->_echo_stderr("Could not notify user about: $message\n");
+        }
 
-            if (isset($this->_prev_bt[$trace]) && array_diff($this->_prev_bt[$trace], $bt)) {
-                $d1 = array_diff($this->_prev_bt[$trace], $bt);
-                $d2 = array_diff($bt, $this->_prev_bt[$trace]);
+        return $this;
+    }
 
-                $this->debug('Call path for '.$trace.' has diverged (was '.implode(', ', $d1).', now '.implode(', ', $d2).")\n");
+    /**
+     * Method designed to intercept one of the hardest-to-debug situations within Agile Toolkit.
+     *
+     * Suppose you define a hook and the hook needs to be called only once, but somehow it is
+     * being called multiple times. You want to know where and how those calls come through.
+     *
+     * Place debugTraceChange inside your hook and give unique $trace identifier. If the method
+     * is invoked through different call paths, this debug info will be logged.
+     *
+     * Do not leave this method in production code !!!
+     *
+     * @param string $trace
+     */
+    public function debugTraceChange($trace = 'default')
+    {
+        $bt = [];
+        foreach (debug_backtrace() as $line) {
+            if (isset($line['file'])) {
+                $bt[] = $line['file'].':'.$line['line'];
             }
+        }
 
-            $this->_prev_bt[$trace] = $bt;
+        if (isset($this->_prev_bt[$trace]) && array_diff($this->_prev_bt[$trace], $bt)) {
+            $d1 = array_diff($this->_prev_bt[$trace], $bt);
+            $d2 = array_diff($bt, $this->_prev_bt[$trace]);
+
+            $this->log('debug', 'Call path for '.$trace.' has diverged (was '.implode(', ', $d1).', now '.implode(', ', $d2).")\n");
```

Comment:
Shouldn't this line better be like below ?
```
$this->debug(true); // force enable debug mode
$this->debug('Call path for '.$trace.' has diverged (was '.implode(', ', $d1).', now '.implode(', ', $d2).")"); // no \n at the end
```

Reply:
log('debug') will work same as debug(), so this way it's simpler.